packages/gatsby-plugin-utils/src/joi.ts

```diff
@@ -0,0 +1,26 @@
+import joi, { Schema } from "joi"
+import { PluginOptionsSchemaJoi } from "./utils/plugin-options-schema-joi-type"
+
+export * from "./utils/plugin-options-schema-joi-type"
+export const Joi: PluginOptionsSchemaJoi = joi.extend({
+  // This tells Joi to extend _all_ types with .dotenv(), see
+  // https://github.com/sideway/joi/commit/03adf22eb1f06c47d1583617093edee3a96b3873
+  // @ts-ignore Joi types weren't updated with that commit, PR: https://github.com/sideway/joi/pull/2477
+  type: /^s/,
+  rules: {
+    dotenv: {
+      args: [`name`],
+      validate(value, helpers, args): void {
+        if (!args.name) {
+          return helpers.error(
+            `any.dotenv requires the environment variable name`
+          )
+        }
+        return value
+      },
+      method(name): Schema {
+        return this.$_addRule({ name: `dotenv`, args: { name } })
+      },
+    },
```

Comment:
I might have forgotten but I'm still not sure why we want this one.

Reply:
This let's plugin authors mark that a value entered during (a potential) `gatsby plugin add` command should be stored in the `.env` file instead of being written to the gatsby-config.js directly. E.g.:

```JS
// gatsby-plugin-source-contentful/gatsby-node.js
export const pluginOptionsSchema = ({ Joi }) => Joi.object({
  accessToken: Joi.string().required().dotenv("CONTENTFUL_ACCESS_TOKEN")
})
```

Then, when we prompt the user to enter their access token we will 1) mask the input in the CLI and 2) write the value to the .env file and `process.env.CONTENTFUL_ACCESS_TOKEN` to the gatsby-config.js 👍

Reply:
It's still something I'm not 100% convinced about. The feature sounds really cool, but I fear that env vars can be similar and bring conflicts.

Can we strip it out until we have the need for it/more discussions?

Reply:
Isn't this the main way we can ensure core -> cloud compatibility? @wardpeet can you explain more about what conflicts you anticipate?

Reply:
I might not 100% understand it so let's say that I have plugin A that has `dotenv('HOST')` and a plugin B `dotenv('HOST')` how would this play out? Also what if for some reason I rather have CONTENTFUL_DELIVERY_TOKEN  instead of CONTENTFUL_ACCESS_TOKEN?

Reply:
So this isn't for the user to understand, but rather the plugin author. We could solve the duplicate problem by concatenating the plugin name and the option name. So `GATSBY_PLUGIN_CONTENTFUL_HOST`. 

The core <-> compatibility is the idea that you'll always be in a valid state and you'd be ready to set up with cloud since it leverages process.env to hook into CMS keys and the like.

Reply:
That's why I want to hold off on adding this feature cause it needs more thinking.

Reply:
I have disabled (i.e. commented out) this extension for now until we decide on a fixed API for it to unblock this PR 👍